### ISSUE TITLE
Gradient compilation takes a significant amount of time

### ISSUE THREAD

**Issue by Jashcraf** [UCSB-Exoplanet-Polarimetry-Lab/GRaTeR-JAX#137]
Used Viztracer to profile the scipy_bounded_optimize and got the following

<img width="1870" height="490" alt="Image" src="https://github.com/user-attachments/assets/1e9303c4-e01a-4f4c-af39-e707faac9daf" />

For those not familiar with reading these kinds of flamegraph, the x-axis is time, and y-axis is function call. This took about 13 minutes to run on an RTX 4090, and 11.5 minutes of that was just compiling the gradient of the objective function.

Not totally sure why - so I thought I'd make an issue while I experiment. It's possible that this is due to either a large number of function calls in the forward model, or because the error in the gradient calculation remains. The L-BFGS-B Solver terminates prematurely, below is the error msg

```python
Bad direction in the line search;
   refresh the lbfgs memory and restart the iteration.

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
   14      5     54     29     0    14   1.800D+02   3.875D+08
  F =   387469377.86612034     

ABNORMAL_TERMINATION_IN_LNSRCH                              

 Line search cannot locate an adequate point after MAXLS
  function and gradient evaluations.
  Previous x, f and g restored.
 Possible causes: 1 error in function or gradient evaluation;
                  2 rounding error dominate computation.
  message: ABNORMAL_TERMINATION_IN_LNSRCH
  success: False
   status: 2
      fun: 72000000000000.0
        x: [ 5.492e+00 -4.832e+00 ... -2.009e+00  0.000e+00]
      nit: 4
      jac: [-2.162e+05  4.509e+01 ...  1.631e+04  1.381e+06]
     nfev: 54
     njev: 54
 hess_inv: <14x14 LbfgsInvHessProduct with dtype=float64>

```
The result of the optimization is still the following, which suggests there isn't a disk lol
<img width="1602" height="633" alt="Image" src="https://github.com/user-attachments/assets/c69fd697-65a2-490d-b970-dc4a00d1175e" />

**Comment by Jashcraf** [comment]
This is the code I'm running to produce the plot above

```python
from astropy.io import fits
import matplotlib.pyplot as plt
import jax
import jax.numpy as jnp
import numpy as np
import pandas as pd
import sys

"""Update compooter params
"""
sys.path.append("/home/jarenashcraft/GRaTeR-JAX/")
jax.config.update("jax_enable_x64", True)

# Grater Jax
from grater_jax.optimization.optimize_framework import Optimizer, OptimizeUtils
from grater_jax.disk_model.objective_functions import (
            objective_model,
            objective_ll,
            objective_fit,
            Parameter_Index
)
from grater_jax.disk_model.SLD_ojax import ScatteredLightDisk
from grater_jax.disk_model.SLD_utils import *

MIN = 80
MAX = 200

if __name__ == "__main__":
    data = fits.getdata("hd117214_H_pol.fits")
    print(data.shape)

    titles = [r'$I$',r'$Q_{\phi}$',r'$U_{\phi}$',]
    plt.figure()
    plt.suptitle("HD 117214 H-band")
    for i, title in enumerate(titles):
        plt.subplot(1, 3, i+1)
        plt.title(title)
        plt.imshow(data[i], cmap="cividis", vmin=0)
        plt.colorbar()




    target_image = OptimizeUtils.process_image(data[1], bounds=(MIN, MAX, MIN, MAX))
    err_map = OptimizeUtils.process_image(data[2], bounds=(MIN, MAX, MIN, MAX))

    # Import Empirical PSF
    scale_factor = 1
    image = fits.open("/home/jarenashcraft/GRaTeR-JAX/PSFs/GPI_Hband_PSF.fits")[0].data[0,:,:]
    scaled_image = (image[::scale_factor, ::scale_factor])[1::, 1::]

    cen_x, cen_y = image.shape[0] // 2, image.shape[1] // 2
    halfmax = (MAX-MIN) // 2

    cropped_image = image[cen_x-halfmax:cen_x+halfmax,
                          cen_y-halfmax:cen_y+halfmax]

    plt.figure()
    plt.title("PSF Image")
    plt.imshow(cropped_image)
    plt.colorbar()

    cropped_image = cropped_image.astype(np.float32)
    def safe_float32_conversion(value):
        try:
            return np.float32(value)
        except (ValueError, TypeError):

            print("This value is unjaxable: " + str(value))

    emp_psf_image = np.nan_to_num(cropped_image)
    emp_psf_image = np.vectorize(safe_float32_conversion)(emp_psf_image)

    start_disk_params = Parameter_Index.disk_params.copy()
    start_spf_params = InterpolatedUnivariateSpline_SPF.params.copy()
    start_psf_params = EMP_PSF.params.copy()
    start_stellar_psf_params = LinearStellarPSF.params.copy()
    start_misc_params = Parameter_Index.misc_params.copy()

    start_spf_params['num_knots'] = len(start_spf_params['knot_values'])

    start_misc_params['nx'] = target_image.shape[0]
    start_misc_params['ny'] = target_image.shape[1]

    start_disk_params['x_center'] = start_misc_params['nx'] // 2
    start_disk_params['y_center'] = start_misc_params['ny'] // 2



    opt = Optimizer(ScatteredLightDisk,
                    DustEllipticalDistribution2PowerLaws,
                    InterpolatedUnivariateSpline_SPF,
                    EMP_PSF,
                    start_disk_params,
                    start_spf_params,
                    start_psf_params,
                    start_misc_params)

    opt.set_empirical_psf(emp_psf_image) # <-- Only necessary if you have an empirical_psf_image
    opt.initialize_knots(target_image) # <-- This needs to be called first before any modeling or jit compiling if spline spf is being used

    opt.jit_compile_model() # <-- opt.log_likelihood is also jit compiled by this
    opt.jit_compile_gradient(target_image, err_map) # <-- Not needed if gradient isn't going to be used
    print(opt.log_likelihood(target_image, err_map))
    opt.get_gradient(['sma', 'alpha_in', 'knot_values'], target_image, err_map)



    plt.imshow(opt.get_model(), origin='lower', cmap='inferno')
    plt.title("Initial Guess, Log-Likelihood: " + str(opt.log_likelihood(target_image, err_map)))



    fit_keys = ['alpha_in', 'alpha_out', 'sma', 'inclination', 'position_angle', 'x_center', 'y_center', 'knot_values', 'e']
    logscaled_params = ['knot_values']
    array_params = ['knot_values']
    knots_shape = np.shape(opt.spf_params["knot_values"])
    bounds = ([0.1, -15, 0, 0, 0, 90, 90, np.zeros(knots_shape), 0],
              [20, -0.1, 150, 180, 400, 110, 110, 1e4*np.ones(knots_shape), 1])

    opt.scipy_bounded_optimize(fit_keys, bounds, logscaled_params, array_params, target_image, err_map, use_grad = True, disp_soln = True, iters = 2000)
    optimal_image = opt.get_model()

    fig, axes = plt.subplots(1,3, figsize=(20,10))

    mask = OptimizeUtils.get_mask(target_image)
    vmin = np.nanpercentile(target_image[mask], 10)
    vmax = np.nanpercentile(target_image[mask], 90)

    im = axes[0].imshow(target_image, origin='lower', cmap='inferno')
    axes[0].set_title("Original Image")
    plt.colorbar(im, ax=axes[0], shrink=0.75)
    im.set_clim(vmin, vmax)

    im = axes[1].imshow(optimal_image, origin='lower', cmap='inferno')
    axes[1].set_title("Disk Fit")
    plt.colorbar(im, ax=axes[1], shrink=0.75)
    im.set_clim(vmin, vmax)

    im = axes[2].imshow(target_image-optimal_image, origin='lower', cmap='inferno')
    axes[2].set_title("Residual")
    plt.colorbar(im, ax=axes[2], shrink=0.75)
    im.set_clim(vmin, vmax)


    plt.show()

```

**Comment by Jashcraf** [comment]
@briley-lewis do you have a good sense of what some good "starting parameters" are for this disk?

**Comment by mihir-r-kondapalli** [comment]
I figured out the issue with the fitting, the error map definition should be:

err_map = OptimizeUtils.process_image(data[2], bounds=(MIN, MAX, MIN, MAX))
instead of:
err_map = OptimizeUtils.process_image(OptimizeUtils.create_empirical_err_map(data[2]), bounds=(MIN, MAX, MIN, MAX))
